### PR TITLE
Purchase order view Take 2

### PR DESCRIPTION
**Pull request by freshteapot**
Brings in #113, #115 and as a fake way to fake the flow into "view" screen of config, not the setup screen

## Summary
- Fake data flow on "view" screen (http://localhost:9007/selfservice/microservices/application/11b6cf47-5d9f-438f-8116-0d9828654657/Dev/view/043a71f8-a7fa-304a-9111-fb2eeea22206?waitForData=fakeit&dataState=waiting) change dataState from "waiting" to "waiting", "flowing", "error".
- Creating a purchaseOrder mircoservice redirects to purchaseOrder view
- The view screen is "first time" aware via query param
- The view screen allows to toggle the edit of the username and password
- Ability to click Generate password for the webhook password
- Fake dataState via the queryParm for the DataStateIcon
- Replaced a few cases of setNotification with the snackbar
- Intoducing "save protection" logic when editing username and password

Linked to
- https://app.asana.com/0/1200899594200551/1201018124798231/f
- https://www.figma.com/file/OPDbO214o4LjpTu316WyMf/Studio-Mockups?node-id=1403%3A72103

### CODE REVIEW

**Review by joelhoisko — CHANGES_REQUESTED** [review]
Nice stuff. Found some places where we can simplify some of the logical operations. I also saw that many of the JSX components had some inconsistent formatting (multiple empty newlines, newlines in weird spots, some components props were multilined while others weren't etc). Maybe we could have some linting to catch these so that it's all nice and consistent (don't have to do it now, but just an observation for us to get the code nicer in the future).

File: Source/SelfService/Web/microservice/create.tsx [hunk 1]
```diff
@@ -106,15 +106,21 @@ export const Create: React.FunctionComponent<Props | undefined> = (props) => {
 
         const kind = searchParams.get('kind') as string;
 
+        if (!kind) {
+            return '';
+        }
+
         if (!items.map(e => e.kind).includes(kind)) {
             return '';
         }
```
[the remark this answers is not in the document]
Comment (freshteapot) [reply Source/SelfService/Web/microservice/create.tsx]: Done.

File: Source/SelfService/Web/microservice/purchaseOrder/configuration.tsx [hunk 2]
```diff
@@ -140,6 +149,30 @@ export const Configuration: React.FunctionComponent<Props> = (props) => {
         }
     };
 
+
+    const allowMicroserviceName = (name): boolean => {
+        return name === '' || name.includes(' ') ? false : true;
```
[the remark this answers is not in the document]
Comment (freshteapot) [reply Source/SelfService/Web/microservice/purchaseOrder/configuration.tsx]: Done

File: Source/SelfService/Web/microservice/purchaseOrder/configuration.tsx [hunk 3]
```diff
@@ -140,6 +149,30 @@ export const Configuration: React.FunctionComponent<Props> = (props) => {
...
+    };
+
+    const allowUsername = (username): boolean => {
+        const cleanedUsername = username.trim();
+        return cleanedUsername === '' || cleanedUsername.includes(' ') ? false : true;
+    };
+
+    const allowPassword = (password): boolean => {
+        return password === '' ? false : true;
```
[the remark this answers is not in the document]
Comment (freshteapot) [reply Source/SelfService/Web/microservice/purchaseOrder/configuration.tsx]: Done

File: Source/SelfService/Web/microservice/purchaseOrder/generatePassword.tsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+export const GeneratePassword: React.FunctionComponent<Props> = (props) => {
+    const { enqueueSnackbar } = useSnackbar();
+    const isCreate = props!.isCreate;
+    const password = props!.password;
+    const setPassword = props!.setPassword;
+    // Using setPassword does not instantly update the password
+    let _password = password;
+
+    // https://app.asana.com/0/1200899594200551/1201018124798232/f
```
[the remark this answers is not in the document]
Comment (freshteapot) [reply Source/SelfService/Web/microservice/purchaseOrder/generatePassword.tsx]: Done

File: Source/SelfService/Web/theme/actionButton.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        },
+    })
+);
+
+
+export const ActionButton: React.FunctionComponent<Props> = (props) => {
+    const classes = useStyles();
+    const _props = props!;
+    const children = _props.children;
+    const onClick = _props.onClick ? _props.onClick : defaultOnClick;
```
[the remark this answers is not in the document]
Comment (freshteapot) [reply Source/SelfService/Web/theme/actionButton.tsx:56]: Done

File: Source/SelfService/Web/utils/httpCredentials.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    return `Basic ${suffix}`;
+};
+
+
+export const makeBearer = (data: ConnectorWebhookConfigBearer): string => {
+    return `Bearer ${data.token}`;
+};
+
+// getCredentialsFromBasicAuth assumes data is "Basic dXNlcjE6aGVsbG93b3JsZA==" where
+// dXNlcjE6aGVsbG93b3JsZA== is user1:helloworld
```
[the remark this answers is not in the document]
Comment (freshteapot) [reply Source/SelfService/Web/utils/httpCredentials.ts]: Done

**Comment by freshteapot** [comment]
I also fixed the defaultOnClick in for SecondaryButton.

I won't lie @joelhoisko the NOR approach does take me a few more seconds to pass when I scan the code, I suspect this will get easier with time.

At the moment, the verbosity of the !NOR ;) feels easier to come back too... onwards.

I will double check I addressed your suggestions and then merge it in.

File: Source/SelfService/Web/microservice/purchaseOrder/configuration.tsx [hunk 7]
```diff
@@ -140,6 +149,30 @@ export const Configuration: React.FunctionComponent<Props> = (props) => {
...
     };
 
+
+    const allowMicroserviceName = (name): boolean => {
+        return name === '' || name.includes(' ') ? false : true;
+    };
+
+    const allowUsername = (username): boolean => {
+        const cleanedUsername = username.trim();
+        return cleanedUsername === '' || cleanedUsername.includes(' ') ? false : true;
```
[the remark this answers is not in the document]
Comment (freshteapot) [reply Source/SelfService/Web/microservice/purchaseOrder/configuration.tsx]: Done

File: Source/SelfService/Web/microservice/purchaseOrder/configuration.tsx [hunk 8]
```diff
@@ -95,29 +95,38 @@ const useStyles = makeStyles((theme: Theme) =>
...
 export const Configuration: React.FunctionComponent<Props> = (props) => {
-    const { setNotification } = useGlobalContext();
     const { enqueueSnackbar } = useSnackbar();
 
     const _props = props!;
     const onSave = _props.onSave;
 
     const ms = _props.microservice;
 
+    const isCreate = ms.name === '' ? true : false;
```
[the remark this answers is not in the document]
Comment (freshteapot) [reply Source/SelfService/Web/microservice/purchaseOrder/configuration.tsx]: Done

File: Source/SelfService/Web/microservice/purchaseOrder/viewConfiguration.tsx [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+    const webhookPrefix = `https://${ms.extra.ingress.host}/api/webhooks`;
+    const webhookPoHead = 'm3/pohead';
+    const webhookPoLine = 'm3/poline';
+    const msName = ms.name;
+
+    useEffect(() => {
+        const firstTime = searchParams.get('firstTime')!;
+        if (firstTime === '1') {
+            console.log('How many times');
```
[the remark this answers is not in the document]
Comment (freshteapot) [reply Source/SelfService/Web/microservice/purchaseOrder/viewConfiguration.tsx:97]: > Let's aim to not have any console.log's in main

A strict man, I agree. Let us use it for the exception, not the norm.

File: Source/SelfService/Web/microservice/purchaseOrder/dataStateIcon.tsx [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+            icon: (<ErrorIcon className={classes.error} />),
+        },
+        waiting: {
+            className: classes.waiting,
+            icon: (<CachedIcon className={classes.waiting} />),
+            text: 'Waiting for data'
+        },
+    };
+
+    const option = options[state] ? options[state] : options.error;
```
[the remark this answers is not in the document]
Comment (freshteapot) [reply Source/SelfService/Web/microservice/purchaseOrder/dataStateIcon.tsx]: Done

**Pull request merged**